File: 1074.number-of-submatrices-that-sum-to-target.cpp
/*
 * @lc app=leetcode id=1074 lang=cpp
 *
 * [1074] Number of Submatrices That Sum to Target
 */

// @lc code=start
#include <vector>
#include <unordered_map>

class Solution {
public:
    int numSubmatrixSumTarget(std::vector<std::vector<int>>& matrix, int target) {
        int m = matrix.size();
        int n = matrix[0].size();

        std::vector<std::vector<int>> prefixSums(m + 1, std::vector<int>(n + 1, 0));
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                prefixSums[i][j] = matrix[i - 1][j - 1];
                if (j > 1) prefixSums[i][j] += prefixSums[i][j - 1];
            }
        }

        int result = 0;
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j <= n; j++) {
                std::unordered_map<int, int> cumuSumSeen;
                int cumuSum = 0;

                for (int row = 1; row <= m; row++) {
                    cumuSum += prefixSums[row][j] - prefixSums[row][i];
                    if (cumuSum == target) {
                        result++;
                    }

                    if (cumuSumSeen.find(cumuSum - target) != cumuSumSeen.end()) {
                        result += cumuSumSeen[cumuSum - target];
                    }

                    cumuSumSeen[cumuSum]++;
                }
            }
        }

        return result;
    }
};

// @lc code=end

File: 1074.number-of-submatrices-that-sum-to-target.py
#
# @lc app=leetcode id=1074 lang=python3
#
# [1074] Number of Submatrices That Sum to Target
#

# @lc code=start
from typing import List


class Solution:
    '''
    2D version of Leetcode 560. Subarray Sum Equals K
    The idea is the same. Calculate prefix sums row by row first
    Then for each columns pair (i,j), add the prefix sum differences row by row
    In each row, ask if new sum - target was seen before    
    '''
    def numSubmatrixSumTarget(self, matrix: List[List[int]], target: int) -> int:
        m, n = len(matrix), len(matrix[0])
        prefix_sums = [[0 for i in range(n + 1)] for j in range(m + 1)]
        for i in range(1, m + 1):
            for j in range(1, n + 1):
                prefix_sums[i][j] = matrix[i - 1][j - 1]
                if j > 1: prefix_sums[i][j] += prefix_sums[i][j - 1]
        
        result = 0
        for i in range(n):
            for j in range(i + 1, n + 1):
                cumu_sum_seen, cumu_sum = {0: 1}, 0
                for row in range(1, m + 1):
                    cumu_sum += prefix_sums[row][j] - prefix_sums[row][i]
                    if cumu_sum - target in cumu_sum_seen:
                        result += cumu_sum_seen[cumu_sum - target]
                    cumu_sum_seen[cumu_sum] = cumu_sum_seen.get(cumu_sum, 0) + 1
        return result
            
# @lc code=end

